### DISCUSSION TITLE
How can we convert list of input devices into a dictionary

### DISCUSSION THREAD

**Discussion by elmcapp** [orchetect/swift-midi#129] [Q&A] [answered]
I am getting the list of inputs using the following

```
let inputDevices = midiManager.endpoints.inputs
print(inputDevices)
```
It simply shows all the midi input devices as a list. 

Does MidiKit provides this same information in a dictionary. If not can you show how to put the device name, and uniqueID in a dictionary

**Comment by orchetect** [comment, marked as answer]
I've added them as properties in MIDIKit 0.8.4.

```swift
midiManager.devices.inputs // returns [MIDIDevice: [MIDIInputEndpoint]]
midiManager.devices.outputs // returns [MIDIDevice: [MIDIOutputEndpoint]]
```

A device can contain zero or more entities, and an entity can contain zero or more inputs and outputs.

These properties just flatten the entities and gives all endpoints associated with a device.
